nsqd/http.go

```diff
@@ -201,10 +208,16 @@ func (s *httpServer) doPUB(w http.ResponseWriter, req *http.Request, ps httprout
 	// add 1 so that it's greater than our max when we test for it
 	// (LimitReader returns a "fake" EOF)
 	readMax := s.nsqd.getOpts().MaxMsgSize + 1
-	body, err := ioutil.ReadAll(io.LimitReader(req.Body, readMax))
+
+	bodyBuffer := s.bodyBufferPoolGet()
+	defer s.bodyBufferPoolPut(bodyBuffer)
+
+	_, err := io.Copy(bodyBuffer, io.LimitReader(req.Body, readMax))
 	if err != nil {
 		return nil, http_api.Err{500, "INTERNAL_ERROR"}
 	}
+
+	body := bodyBuffer.Bytes()
```

Comment:
I think there is an issue with lifetimes.

The `bodyBuffer` is put back to the pool at the end of the function. This `body` is a byte slice which is a reference into the buffer owned by the `bytes.Buffer`, which is reused when the `bytes.Buffer` is reset and reused. Below, `NewMessage()` stores the byte slice (reference) in the Message. That message is put to the Topic, where it it can be retained in the memoryMsgChan for some time (usually a short amount but not always), and when the message is copied to the Channels the body byte slice still references the original buffer owned by the bytes.Buffer. The body contents are never actually copied (unless to/from the "backend disk-queue").

Reply:
Thank you for your guidance. I ignored this point when I used it.

Later, I tried to use a global bufferPool for optimization. When reading request body applies for a buffer and associates it with the message, the buffer is released to the pool after the consumer sends the FIN or when the diskqueue is used, but the effect of the stress test is not ideal and increases the complexity of the code.

Reply:
I just thought about it. There are still some problems with my above design. Because there may be multiple channels and a message may be consumed several times, it is impossible to determine whether to recycle the buffer through the FIN.

I have no idea